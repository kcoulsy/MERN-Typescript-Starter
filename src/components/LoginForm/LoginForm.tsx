import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import Card from '../Card/Card';
import Loader from '../Loader/Loader';

import { startLogin } from '../../actions/auth';

interface LoginFormProps {
    startLogin: Function;
    attemptingLogin: boolean;
    loginAttemptFailed: boolean;
}

interface LoginFormErrorStateObj {
    usernameEmpty?: boolean;
    passwordEmpty?: boolean;
    invalidLogin?: boolean;
}

export const LoginForm: React.FunctionComponent<LoginFormProps> = ({
    startLogin,
    attemptingLogin,
    loginAttemptFailed,
}) => {
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');

    const defaultErrorObj = {
        usernameEmpty: false,
        passwordEmpty: false,
        invalidLogin: false,
    };

    const [error, setError] = useState<LoginFormErrorStateObj>(defaultErrorObj);

    const submitForm = (ev: React.MouseEvent<HTMLButtonElement>) => {
        ev.preventDefault();
        if (!username.length || !password.length) {
            setError({
                usernameEmpty: !username.length,
                passwordEmpty: !password.length,
            });
        } else {
            setError(defaultErrorObj);
            startLogin(username, password);
        }
    };

    const formError = loginAttemptFailed || Object.values(error).includes(true);

    return (
        <Card title="Login" centered>
            <Loader isLoading={attemptingLogin} loadingText="Logging in">
                <form className={`ui form${formError ? ' error' : ''}`}>
                    <div
                        className={`field${
                            error.usernameEmpty ? ' error' : ''
                        }`}>
                        <label>Username</label>
                        <input
                            type="text"
                            name="username"
                            placeholder="Username"
                            onChange={(ev) => setUsername(ev.target.value)}
                        />
                    </div>
                    <div
                        className={`field${
                            error.passwordEmpty ? ' error' : ''
                        }`}>
                        <label>Password</label>
                        <input
                            type="password"
                            name="password"
                            placeholder="Password"
                            onChange={(ev) => setPassword(ev.target.value)}
                        />
                    </div>
                    <button
                        className="ui button"
                        type="submit"
                        onClick={submitForm}>
                        Login
                    </button>
                    Not registered? <Link to="/register">Register</Link>
                    {formError ? (
                        <div className="ui error message">
                            <ul className="list">
                                {error.usernameEmpty ? (
                                    <li>Please enter a username</li>
                                ) : null}
                                {error.passwordEmpty ? (
                                    <li>Please enter a password</li>
                                ) : null}
                                {loginAttemptFailed ? (
                                    <li>Login Invalid</li>
                                ) : null}
                            </ul>
                        </div>
                    ) : null}
                </form>
            </Loader>
        </Card>
    );
};

// TODO ROOT STATE
export const mapStateToProps = (state: { auth: { attemptingLogin: boolean; loginAttemptFailed: boolean; }}) => {
    return {
        attemptingLogin: state.auth.attemptingLogin,
        loginAttemptFailed: state.auth.loginAttemptFailed,
    };
};
const mapDispatchToProps = { startLogin };

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);
