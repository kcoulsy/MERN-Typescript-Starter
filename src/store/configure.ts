import { configureStore } from '@reduxjs/toolkit';

// Import Reducers
import authReducer from '../reducers/auth';

const store = configureStore({
    reducer: {
        // reducers here
        auth: authReducer,
    },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;
