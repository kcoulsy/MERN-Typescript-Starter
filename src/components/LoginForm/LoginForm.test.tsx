import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { LoginForm, mapStateToProps } from './LoginForm';
import { RootState } from '../../store/configure';

describe('render login form component', () => {
    it('should render LoginForm component', () => {
        const startLogin = jest.fn();
        const wrapper = shallow(<LoginForm startLogin={startLogin} attemptingLogin={false} loginAttemptFailed={false} />);
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should call startLogin on submit', () => {
        const startLogin = jest.fn();
        const wrapper = shallow(<LoginForm startLogin={startLogin} attemptingLogin={false} loginAttemptFailed={false} />);
        const testUser = {
            username: 'username',
            password: 'password',
        };
        wrapper.find('input[name="username"]').simulate('change', { target: { value: testUser.username } });
        wrapper.find('input[name="password"]').simulate('change', { target: { value: testUser.password } });
        wrapper.find('button[type="submit"]').simulate('click', { preventDefault: jest.fn() });

        expect(startLogin.mock.calls.length).toBe(1);
        expect(startLogin.mock.calls[0][0]).toBe(testUser.username);
        expect(startLogin.mock.calls[0][1]).toBe(testUser.password);
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('should show errors if field is empty', () => {
        const startLogin = jest.fn();
        const wrapper = shallow(<LoginForm startLogin={startLogin} attemptingLogin={false} loginAttemptFailed={false} />);
        const testUser = {
            username: 'username',
            password: 'password',
        };

        //testing if password is missing, show the password error
        wrapper.find('input[name="username"]').simulate('change', { target: { value: testUser.username } });
        wrapper.find('input[name="password"]').simulate('change', { target: { value: '' } });
        wrapper.find('button[type="submit"]').simulate('click', { preventDefault: jest.fn() });

        expect(startLogin.mock.calls.length).toBe(0);
        expect(wrapper.find('.error.message').length).toBe(1);
        expect(wrapper.find('.error.message').contains('Please enter a password')).toBe(true);

        // typing in that password should remove the error on click
        wrapper.find('input[name="password"]').simulate('change', { target: { value: testUser.password } });
        wrapper.find('button[type="submit"]').simulate('click', { preventDefault: jest.fn() });

        expect(startLogin.mock.calls.length).toBe(1);
        expect(wrapper.find('.error.message').length).toBe(0);

        //////////////////////

        // Now testing if username is missing, show the username error
        wrapper.find('input[name="username"]').simulate('change', { target: { value: '' } });
        wrapper.find('input[name="password"]').simulate('change', { target: { value: testUser.password } });
        wrapper.find('button[type="submit"]').simulate('click', { preventDefault: jest.fn() });

        expect(startLogin.mock.calls.length).toBe(1); // still 1 from testing password
        expect(wrapper.find('.error.message').length).toBe(1);
        expect(wrapper.find('form').hasClass('error')).toBe(true);
        expect(wrapper.find('.error.message').contains('Please enter a username')).toBe(true);

        // typing in that username should remove the error on click
        wrapper.find('input[name="username"]').simulate('change', { target: { value: testUser.username } });
        wrapper.find('button[type="submit"]').simulate('click', { preventDefault: jest.fn() });

        expect(startLogin.mock.calls.length).toBe(2);
        expect(wrapper.find('.error.message').length).toBe(0);
    });

    it('should show error login attempt failed', () => {
        const startLogin = jest.fn();
        const wrapper = shallow(<LoginForm startLogin={startLogin} attemptingLogin={false} loginAttemptFailed={false} />);

        expect(wrapper.find('.error.message').length).toBe(0);

        wrapper.setProps({ loginAttemptFailed: true });
        expect(wrapper.find('.error.message').length).toBe(1);
        expect(wrapper.find('.error.message').contains('Login Invalid')).toBe(true);
    });

    it('should correctly map state to props', () => {
        const mockState: RootState = {
            auth: {
                attemptingLogin: false,
                token: null,
                loginAttemptFailed: false,
                appInitialising: false,
                appInitialised: false,
                attemptingRegister: false,
            },
        };

        expect(mapStateToProps(mockState).attemptingLogin).toEqual(mockState.auth.attemptingLogin);
        expect(mapStateToProps(mockState).loginAttemptFailed).toEqual(mockState.auth.loginAttemptFailed);
    });
});
