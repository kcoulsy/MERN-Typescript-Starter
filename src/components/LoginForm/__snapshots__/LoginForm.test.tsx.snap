// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render login form component should call startLogin on submit 1`] = `
<Card
  centered={true}
  title="Login"
>
  <Loader
    isLoading={false}
    loadingText="Logging in"
  >
    <form
      className="ui form"
    >
      <div
        className="field"
      >
        <label>
          Username
        </label>
        <input
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
        />
      </div>
      <div
        className="field"
      >
        <label>
          Password
        </label>
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
      </div>
      <button
        className="ui button"
        onClick={[Function]}
        type="submit"
      >
        Login
      </button>
      Not registered? 
      <Link
        to="/register"
      >
        Register
      </Link>
    </form>
  </Loader>
</Card>
`;

exports[`render login form component should render LoginForm component 1`] = `
<Card
  centered={true}
  title="Login"
>
  <Loader
    isLoading={false}
    loadingText="Logging in"
  >
    <form
      className="ui form"
    >
      <div
        className="field"
      >
        <label>
          Username
        </label>
        <input
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
        />
      </div>
      <div
        className="field"
      >
        <label>
          Password
        </label>
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
      </div>
      <button
        className="ui button"
        onClick={[Function]}
        type="submit"
      >
        Login
      </button>
      Not registered? 
      <Link
        to="/register"
      >
        Register
      </Link>
    </form>
  </Loader>
</Card>
`;
